File: src/ndi/+ndi/+cloud/+internal/duplicateDocuments.m
function [duplicateDocs, originalDocs] = duplicateDocuments(cloudDatasetId, options)
% DUPLICATEDOCUMENTS - Find and optionally remove duplicate documents in a cloud dataset
%
% [DUPLICATEDOCS, ORIGINALDOCS] = ndi.cloud.internal.duplicateDocuments(CLOUDDATASETID, ...)
%
% This function identifies duplicate documents within a specified NDI cloud dataset.
% Duplicates are defined as documents that share the same 'ndi.document.id' (or 'name'
% as a fallback) but may have different cloud-specific '_id' values.
%
% The function determines which document to keep as the 'original' and which to
% mark as a 'duplicate' based on the alphabetical order of their unique cloud document
% ID ('id'). The one with the alphabetically earliest 'id' is kept.
%
% By default, this function will delete the identified duplicate documents from the
% cloud dataset. This behavior can be controlled using a name-value pair.
%
% This function can take additional name-value pair arguments:
% |--------------------------|-----------------------------------------------|
% | 'deleteDuplicates'       | A boolean value (true/false) indicating       |
% |                          | whether to delete the identified duplicates.  |
% |                          | (Default: true)                               |
% | 'maximumDeleteBatchSize' | The maximum number of documents to delete in  |
% |                          | a single bulk operation.                      |
% |                          | (Default: 100)                                |
% | 'verbose'                | A boolean value (true/false) indicating       |
% |                          | whether to display status messages.           |
% |                          | (Default: false)                              |
%
% Outputs:
%   DUPLICATEDOCS - A struct array of the documents that were identified as
%                   duplicates. These are the documents that were deleted if
%                   'deleteDuplicates' is true.
%   ORIGINALDOCS  - A struct array of the documents that were identified as
%                   the originals.
%
% See also: ndi.cloud.api.documents.listDatasetDocumentsAll,
%           ndi.cloud.api.documents.bulkDeleteDocuments

arguments
    cloudDatasetId (1,:) char {mustBeText}
    options.deleteDuplicates (1,1) logical = true
    options.maximumDeleteBatchSize (1,1) {mustBeInteger, mustBePositive} = 100
    options.verbose (1,1) logical = false
end

duplicateDocs = struct([]);
originalDocs = struct([]);

if options.verbose
    disp(['Searching for all documents...']);
end
[~, allDocsStruct] = ndi.cloud.api.documents.listDatasetDocumentsAll(cloudDatasetId);
if options.verbose
    disp(['Done.']);
end

if isempty(allDocsStruct)
    return;
end

% --- Struct Normalization ---
% Get all unique field names from all document summaries
allFields = {};
for i = 1:numel(allDocsStruct)
    allFields = union(allFields, fieldnames(allDocsStruct(i)));
end

% Create a template struct with all fields initialized to empty
templateStruct = struct();
for i = 1:numel(allFields)
    templateStruct.(allFields{i}) = [];
end

% Create a normalized array of structs
allDocs = repmat(templateStruct, numel(allDocsStruct), 1);
for i = 1:numel(allDocsStruct)
    currentDoc = allDocsStruct(i);
    fields = fieldnames(currentDoc);
    for j = 1:numel(fields)
        allDocs(i).(fields{j}) = currentDoc.(fields{j});
    end
end
% --- End Normalization ---

% Now that we have a homogeneous struct array, initialize the outputs
duplicateDocs = allDocs([]);

docMap = containers.Map('KeyType', 'char', 'ValueType', 'any');

% Identify originals and duplicates
for i = 1:numel(allDocs)
    currentDoc = allDocs(i);

    if isfield(currentDoc, 'ndiId') && ~isempty(currentDoc.ndiId)
        doc_group_key = currentDoc.ndiId;
    else
        doc_group_key = currentDoc.name;
    end

    if ~isKey(docMap, doc_group_key)
        docMap(doc_group_key) = currentDoc;
    else
        existingDoc = docMap(doc_group_key);
        if strcmp(currentDoc.id, existingDoc.id) < 0
            duplicateDocs(end+1) = existingDoc;
            docMap(doc_group_key) = currentDoc;
        else
            duplicateDocs(end+1) = currentDoc;
        end
    end
end

originalDocsCell = values(docMap);
if ~isempty(originalDocsCell)
    originalDocs = [originalDocsCell{:}]';
else
    originalDocs = allDocs([]);
end

if options.deleteDuplicates && ~isempty(duplicateDocs)
    if options.verbose
        disp(['Found ' int2str(numel(duplicateDocs)) ' duplicates to delete.']);
    end

    docIdsToDelete = {duplicateDocs.id};

    numBatches = ceil(numel(docIdsToDelete) / options.maximumDeleteBatchSize);

    for i = 1:numBatches
        startIndex = (i-1) * options.maximumDeleteBatchSize + 1;
        endIndex = min(i * options.maximumDeleteBatchSize, numel(docIdsToDelete));

        batchIds = docIdsToDelete(startIndex:endIndex);

        if options.verbose
            disp(['Deleting batch ' int2str(i) ' of ' int2str(numBatches) '...']);
        end
        ndi.cloud.api.documents.bulkDeleteDocuments(cloudDatasetId, batchIds);
        if options.verbose
            disp(['Batch ' int2str(i) ' deleted.']);
        end
    end

    if options.verbose
        disp('All duplicate documents deleted.');
    end
else
    if isempty(duplicateDocs)
        if options.verbose
            disp('No duplicate documents found.');
        end
    else
        if options.verbose
            disp(['Found ' int2str(numel(duplicateDocs)) ' duplicates, but deletion was not requested.']);
        end
    end
end

end

File: ndi-python/ndi/cloud/internal/duplicate_documents.py
"""
Find and optionally remove duplicate documents in a cloud dataset.

This module identifies documents with the same NDI ID but different cloud IDs,
and optionally deletes the duplicates.

MATLAB Source: ndi/+ndi/+cloud/+internal/duplicateDocuments.m
"""

from typing import Tuple, List, Dict, Any
import math


def duplicate_documents(
    cloud_dataset_id: str,
    delete_duplicates: bool = True,
    maximum_delete_batch_size: int = 1000,
    verbose: bool = False
) -> Tuple[List[Dict[str, Any]], List[Dict[str, Any]]]:
    """
    Find and optionally remove duplicate documents in a cloud dataset.

    This function identifies duplicate documents within a specified NDI cloud dataset.
    Duplicates are defined as documents that share the same 'ndi.document.id' (or 'name'
    as a fallback) but may have different cloud-specific '_id' values.

    The function determines which document to keep as the 'original' and which to
    mark as a 'duplicate' based on the alphabetical order of their unique cloud document
    ID ('id'). The one with the alphabetically earliest 'id' is kept.

    By default, this function will delete the identified duplicate documents from the
    cloud dataset. This behavior can be controlled using a parameter.

    Args:
        cloud_dataset_id: The unique identifier of the cloud dataset
        delete_duplicates: If True, deletes the identified duplicates. Default: True
        maximum_delete_batch_size: The maximum number of documents to delete in
            a single bulk operation. Default: 1000
        verbose: If True, displays status messages. Default: False

    Returns:
        Tuple containing:
            - duplicate_docs: List of dictionaries representing documents that were
              identified as duplicates. These are the documents that were deleted
              if delete_duplicates is True.
            - original_docs: List of dictionaries representing documents that were
              identified as the originals.

    Raises:
        RuntimeError: If the API call to list documents fails

    Example:
        >>> from ndi.cloud.internal import duplicate_documents
        >>> # Find and delete duplicates
        >>> dups, originals = duplicate_documents('d-12345', verbose=True)
        >>> print(f'Found {len(dups)} duplicates and {len(originals)} originals')
        >>>
        >>> # Find but don't delete duplicates
        >>> dups, originals = duplicate_documents('d-12345', delete_duplicates=False)

    See Also:
        ndi.cloud.api.documents.list_dataset_documents_all
        ndi.cloud.api.documents.bulk_delete_documents

    MATLAB Source Reference:
        ndi/+ndi/+cloud/+internal/duplicateDocuments.m
    """
    from ndi.cloud.api.documents import list_dataset_documents_all, bulk_delete_documents

    duplicate_docs = []
    original_docs = []

    if verbose:
        print('Searching for all documents...')

    success, all_docs_list = list_dataset_documents_all(cloud_dataset_id)
    if not success:
        raise RuntimeError(f'Failed to list dataset documents: {all_docs_list}')

    if verbose:
        print('Done.')

    if not all_docs_list:
        return duplicate_docs, original_docs

    # Create a dictionary to track originals by their NDI ID or name
    doc_map = {}

    # Identify originals and duplicates
    for current_doc in all_docs_list:
        # Use ndiId if available, otherwise use name as the grouping key
        if 'ndiId' in current_doc and current_doc['ndiId']:
            doc_group_key = current_doc['ndiId']
        else:
            doc_group_key = current_doc.get('name', '')

        if not doc_group_key:
            # Skip documents without identifiers
            continue

        if doc_group_key not in doc_map:
            # First occurrence - mark as original
            doc_map[doc_group_key] = current_doc
        else:
            # Found a duplicate
            existing_doc = doc_map[doc_group_key]

            # Compare cloud IDs alphabetically to determine which to keep
            # The one with the smaller (earlier) ID is kept as original
            if current_doc['id'] < existing_doc['id']:
                # Current doc has earlier ID, so it becomes the original
                duplicate_docs.append(existing_doc)
                doc_map[doc_group_key] = current_doc
            else:
                # Existing doc remains original, current is duplicate
                duplicate_docs.append(current_doc)

    # Extract original documents from the map
    original_docs = list(doc_map.values())

    # Delete duplicates if requested
    if delete_duplicates and duplicate_docs:
        if verbose:
            print(f'Found {len(duplicate_docs)} duplicates to delete.')

        # Extract IDs of documents to delete
        doc_ids_to_delete = [doc['id'] for doc in duplicate_docs]

        # Calculate number of batches needed
        num_batches = math.ceil(len(doc_ids_to_delete) / maximum_delete_batch_size)

        # Delete in batches
        for i in range(num_batches):
            start_index = i * maximum_delete_batch_size
            end_index = min((i + 1) * maximum_delete_batch_size, len(doc_ids_to_delete))

            batch_ids = doc_ids_to_delete[start_index:end_index]

            if verbose:
                print(f'Deleting batch {i + 1} of {num_batches}...')

            bulk_delete_documents(cloud_dataset_id, batch_ids)

            if verbose:
                print(f'Batch {i + 1} deleted.')

        if verbose:
            print('All duplicate documents deleted.')
    else:
        if not duplicate_docs:
            if verbose:
                print('No duplicate documents found.')
        else:
            if verbose:
                print(f'Found {len(duplicate_docs)} duplicates, but deletion was not requested.')

    return duplicate_docs, original_docs
